content/BPIP-3.md

```diff
@@ -146,28 +149,113 @@ The following method is added.
+- preminted vouchers, with no exchange associated and
+- true vouchers, with exactly one exchange associated.
+  
+In current protocol, all vouchers are issued when protocol invokes `issueVoucher` method. In this proposal we add additional method `preMint` which allows seller (operator) to issue a desired number of vouchers.
+Since currently it holds that exchange id (in protocol) always matches token id (in voucher contract), it's desired that this stays even with premint. To achieve it, we propose a method to reserve range (`reserveRange`), which effectively reserves a desired number of exchange ids in the protocol. Then whenever a preminted voucher is converted into a true voucher, it keeps its token id and gets matching exchange id in the protocol.
+
+Although range reservation ultimately happens on voucher contract, method must be invoked true the protocol. That way it can be ensured that the amount of preminted vouchers never exceeds quantity available, set in the offer.
+
+If voucher expires or is voided, it currently means that buyers cannot commit to it anymore. However, with preminted vouchers it's possible that some of them were not bought before that happen. In this case protocol should also prevent committing to the offer, which then means that some of unusable vouchers are stuck in the contract and possibly on marketplaces. To delist them we suggest method to burn all preminted vouchers (`burnPremintedVouchers`) that cannot be redeemed anymore.
+
+Given that now buyer does not directly interact with the protocol, this opens the question, how to ensure protocol exchange guarantees (i.e. possibility to get funds back if something goes wrong). The proposed solution is than now seller needs to provide enough funds to cover both item price and seller deposit. This ensures buyer protection straight away, but it increases a seller's capital requirement a bit. However, this drawback is not to big, since as soon as marketplace exchange happens, seller receives the funds, which can be directly used to cover item price in next sales. Moreover, as soon as some exchange are finalized and funds are released back into seller's pool, those funds can be automatically used to cover future sales.
+
+This proposal might rise some efficiency concerns about the solution efficiency, since minting ERC721 is relatively costly operation. ERC1155 was considered as alternative, but it turned out that with it it's impossible to maintain nice exchange id and token id equivalence. To overcome the problem of ERC721 inefficiency we propose another solution. Since premint always assign token ownership to voucher contract owner, there is no need to explicitly store it's address during the premint. It's enough that only correct transfer events are emitted, while general consistency is achieved by overriding ERC721 `ownerOf`, `transferFrom` and `safeTransferFrom` methods to correctly manage all tokens that were preminted.
```

Comment:
```suggestion
This proposal might raise some efficiency concerns about the solution efficiency since minting ERC721 is a relatively costly operation. ERC1155 was considered as an alternative, but it turned out that with it it's impossible to maintain a nice exchange id and token id equivalence. To overcome the problem of ERC721 inefficiency, we propose another solution. Since premint always assigns token ownership to the voucher contract owner, there is no need to explicitly store its address during the premint. It's enough that only correct transfer events are emitted, while general consistency is achieved by overriding ERC721 `ownerOf`, `transferFrom` and `safeTransferFrom` methods to correctly manage all tokens that were preminted.
```